lightning/src/ln/async_payments_tests.rs

```diff
@@ -8,33 +8,100 @@
 // licenses.
 
 use crate::blinded_path::message::{MessageContext, OffersContext};
-use crate::events::{Event, HTLCDestination, MessageSendEventsProvider, PaymentFailureReason};
-use crate::ln::blinded_payment_tests::get_blinded_route_parameters;
+use crate::blinded_path::payment::PaymentContext;
+use crate::blinded_path::payment::{AsyncBolt12OfferContext, BlindedPaymentTlvs};
+use crate::chain::channelmonitor::{HTLC_FAIL_BACK_BUFFER, LATENCY_GRACE_PERIOD_BLOCKS};
+use crate::events::{
+	Event, HTLCDestination, MessageSendEvent, MessageSendEventsProvider, PaymentFailureReason,
+};
+use crate::ln::blinded_payment_tests::{fail_blinded_htlc_backwards, get_blinded_route_parameters};
 use crate::ln::channelmanager::{PaymentId, RecipientOnionFields};
 use crate::ln::functional_test_utils::*;
+use crate::ln::msgs;
 use crate::ln::msgs::ChannelMessageHandler;
 use crate::ln::msgs::OnionMessageHandler;
 use crate::ln::offers_tests;
 use crate::ln::onion_utils::INVALID_ONION_BLINDING;
+use crate::ln::outbound_payment::PendingOutboundPayment;
 use crate::ln::outbound_payment::Retry;
+use crate::offers::invoice_request::InvoiceRequest;
 use crate::offers::nonce::Nonce;
 use crate::offers::offer::Offer;
 use crate::offers::static_invoice::StaticInvoice;
-use crate::onion_message::async_payments::{
-	AsyncPaymentsMessage, AsyncPaymentsMessageHandler, ReleaseHeldHtlc,
-};
+use crate::onion_message::async_payments::{AsyncPaymentsMessage, AsyncPaymentsMessageHandler};
 use crate::onion_message::messenger::{Destination, MessageRouter, MessageSendInstructions};
 use crate::onion_message::offers::OffersMessage;
 use crate::onion_message::packet::ParsedOnionMessageContents;
 use crate::prelude::*;
+use crate::routing::router::{Payee, PaymentParameters};
+use crate::sign::NodeSigner;
+use crate::sync::Mutex;
 use crate::types::features::Bolt12InvoiceFeatures;
-use crate::types::payment::{PaymentPreimage, PaymentSecret};
+use crate::types::payment::{PaymentHash, PaymentPreimage, PaymentSecret};
+use bitcoin::constants::ChainHash;
+use bitcoin::network::Network;
 use bitcoin::secp256k1;
 use bitcoin::secp256k1::Secp256k1;
 
 use core::convert::Infallible;
 use core::time::Duration;
 
+// Goes through the async receive onion message flow, returning the final release_held_htlc OM.
```

Comment:
For reviewers: the last commit here is rather large here but it's all tests and test util changes. Let me know if it would be helpful to break it up!